### PR TITLE
feat(qwen3): multi-position hedged verification for the DSpark lane

### PR DESCRIPTION
**Pull request by FeathBow**
## Description

A third of DSpark's speculative rounds accept zero draft tokens on agentic traffic — a full target forward buys one token — yet at the exact position where the chain breaks, the Markov head's runner-up is the committed token 29.6% of the time. Single-chain decoding throws that capacity away. This change verifies it instead: each draft round also proposes one alternative chain per configured branch position (`PEGAINFER_SPEC_HEDGE=<H>`, `PEGAINFER_SPEC_HEDGE_POSITIONS=0,1,2`) — the exact Markov top-2 at the branch, taken from a device-side snapshot, with greedy continuation over the same backbone logits — and all chains ride the existing verify pass as extra spans on the bucket-aligned CUDA graphs. The longest-matching chain wins and its scratch KV pages copy back; greedy output stays byte-identical, and the KV transaction interface and the draft/verify token-span seam are unchanged. Branch positions come from an offline counterfactual study over 10,801 logged break events: branching at the next position outprices deeper candidates at the same position.

Hedge eligibility travels the draft seam as a remaining-output budget: only a greedy request whose earliest branch position survives the verify-side span clamp can consume a hedge slot, and draft and verify select from the same list, so batch order never changes whether hedging runs. An opt-in controller (`PEGAINFER_SPEC_HEDGE_AUTO=1`) picks the chain count at runtime by measured tokens/sec — median window rate with an ascending margin, so flat traffic settles at zero chains. Kernel surface: a two-stage strided `markov_step_top2`, `hedge_ladder_force`, and a `chains` divisor plus an optional request map on the markov step kernels (the mapped variants are `unsafe` with documented device-index preconditions). KV surface: scratch pages outside the allocatable pool (`new_with_scratch_pages`) and a raw `copy_page` primitive.

### Test Env

- Single GPU (48 GB, sm_89, x86_64), CUDA 12.9.
- Qwen3-4B target + DSpark block7 drafter checkpoint; agentic coding traces (12.4-12.7k-token prompts), greedy, bs=1.

### Verification

- fmt and the crate unit suite green (97 tests, including order-invariance and span-clamp regressions for hedge selection and the controller's executed-configuration bookkeeping); `cargo check -p pegainfer-k3` green on the shared-kernel signature change. glm52's two identical call-site updates do not compile on this host (its kernels require sm_100+) — please re-verify in a cu132 environment.
- `dflash_speculative_gate` 5/5 on the exact base `bdbce8f4`, with `PEGAINFER_REQUIRE_HEDGE_GATE=1`: the hedged execution gate re-runs the sequential and concurrent-heterogeneous lossless suites in a child process with the hedge enabled, and its pass requires parsed per-round counters — hedged rounds > 0, at least one chain win (exercising winner copy-back and hidden-row compaction), and at least one loss round. The draft-headroom case is fixed in this PR (constants read from both checkpoint configs; unconstructable pairings skip explicitly).
- Wall-clock A/B, five arms x five interleaved 40-prompt chunks on the same 200 prompts, quiet host, output byte-identical across arms: baseline 572s; +1 chain 565s (-1.2%); +2 chains 555s (-3.0%); +3 chains 549s (-4.0%); +4 chains 553s (-3.3%). Within-arm chunk spread is <=2s and C=3 beats C=4 in all five paired chunks — the knee is systemic (five verify spans fall off the CUDA-graph batch buckets into eager execution), not yield saturation.
- Accept-length probes on the same prompts (9.2k-11k rounds per arm): committed tokens per verify round 2.302 -> 2.744 at three chains (+19.2%), verify rounds -16.1%; the fourth chain's marginal yield lands exactly where the offline study priced it (+0.027 measured vs +0.028 predicted). Easy traffic (HumanEval) is neutral and the auto controller settles at C=0 there.

### CODE REVIEW

File: pegainfer-qwen3/src/executor/dflash_lane.rs [hunk 1]
```diff
@@ -278,8 +333,76 @@ impl LocalQwen3Lane {
...
+            // hedge: eligibility travels on the draft item, and selection is
+            // capped by the per-round hedge cap and the ladder's scratch
+            // budget at the chain count that actually runs this round.
+            let hedged = select_hedged_requests(
+                requests,
+                *hedge_cap,
+                match ladder_chains {
+                    0 => 0,
+                    c => scratch.markov_max_batch() / c,
+                },
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply pegainfer-qwen3/src/executor/dflash_lane.rs:406]: Good catch

**Comment by FeathBow** [comment]
@codex check again please

File: pegainfer-qwen3/src/executor/auto_hedge.rs [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+    /// the round actually ran — 0 when a guard disabled hedging, whatever
+    /// [`Self::current_c`] requested.
+    pub(super) fn tick(&mut self, executed_c: usize, committed_tokens: usize) {
+        let now = Instant::now();
+        let Some(prev) = self.last_tick.replace(now) else {
+            return;
+        };
+        let elapsed = now - prev;
+        if elapsed > ROUND_GAP_CUTOFF || executed_c >= self.stats.len() {
+            return;
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply pegainfer-qwen3/src/executor/auto_hedge.rs]: the long gap is now fixed by cutting off the active window

File: pegainfer-qwen3/src/executor/dflash_lane.rs [hunk 3]
```diff
@@ -338,3 +468,175 @@ impl LocalQwen3Lane {
...
+}
+
+/// `PEGAINFER_SPEC_HEDGE=<H>` enables the parallel hedge and caps the number
+/// of hedged requests per verify round at `H` (each hedged request costs one
+/// extra verify span and two lane scratch KV pages per hedge chain). `0`/
+/// unset/unparsable = off. Read once.
+pub(super) fn spec_hedge_cap() -> usize {
+    static CAP: std::sync::OnceLock<usize> = std::sync::OnceLock::new();
+    *CAP.get_or_init(|| {
+        std::env::var("PEGAINFER_SPEC_HEDGE")
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply pegainfer-qwen3/src/executor/dflash_lane.rs:555]: Sure

File: docs/models/qwen3/dspark-integration.md [hunk 4]
```diff
@@ -1,6 +1,6 @@
...
...rkov_w1 [V,256]`, `markov_w2 [V,256]`, `confidence_head.proj.weight [1,2816]`, `.bias [1]`), and an "anchor-is-the-first-prediction" block-layout tweak. **Our verify/accept/KV-transaction stack is reused unchanged** — DSpark only changes the *propose* step (the later env-gated hedge, `PEGAINFER_SPEC_HEDGE*`, additionally expands verify with extra chain spans over scratch KV pages; off by default). On the 5090 greedy sweep, DSpark beats the matched block7 DFlash baseline by **+3.6% geomean output tok/s** overall, with the expected wins on text/code (+3.1% to +15.8%; random is the synthetic exception) and a better accepted-draft distribution (**2.52 vs 2.30** accepted draft tokens/round, full-7 accept **17.4% vs 13.9%**). Decisions settled: **block_size = 7**, **reuse the target head** (DSpark's `embed_tokens`/`lm_head` are byte-identical to target Qwen3-4B's — skip loading them). **Phase 1 delta over DFlash is small: config + 2 tensor loads + one sequential Markov loop in `execute_dflash_draft`; no new verify/KV/graph** *(true of Phase 1 as landed — the later opt-in hedge adds expanded verify spans, scratch KV pages, and per-shape Markov graphs, all behind `PEGAINFER_SPEC_HEDGE*`)*.
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply docs/models/qwen3/dspark-integration.md:3]: fine
